File: examples/leetcode-out/fortran/3/longest-substring-without-repeating-characters.f90
program main
  implicit none
  call test_example_1()
  call test_example_2()
  call test_example_3()
  call test_empty_string()
contains
  function lengthOfLongestSubstring(s) result(res)
    implicit none
    integer(kind=8) :: res
    character(len=*), intent(in) :: s
    integer(kind=8) :: best
    integer(kind=8) :: i
    integer(kind=8) :: j
    integer(kind=8) :: length
    integer(kind=8) :: n
    integer(kind=8) :: start
    n = len(s)
    start = 0_8
    best = 0_8
    i = 0_8
    do while ((i < n))
      j = start
      do while ((j < i))
        if ((s(modulo(j, len(s)) + 1:modulo(j, len(s)) + 1) == s(modulo(i, len(s)) + 1:modulo(i, len(s)) + 1))) then
          start = (j + 1_8)
          exit
        end if
        j = (j + 1_8)
      end do
      length = ((i - start) + 1_8)
      if ((length > best)) then
        best = length
      end if
      i = (i + 1_8)
    end do
    res = best
    return
  end function lengthOfLongestSubstring

  subroutine test_example_1()
    implicit none
    if (.not. (lengthOfLongestSubstring('abcabcbb') == 3_8)) then
      print *, 'expect failed'
      stop 1
    end if
  end subroutine test_example_1

  subroutine test_example_2()
    implicit none
    if (.not. (lengthOfLongestSubstring('bbbbb') == 1_8)) then
      print *, 'expect failed'
      stop 1
    end if
  end subroutine test_example_2

  subroutine test_example_3()
    implicit none
    if (.not. (lengthOfLongestSubstring('pwwkew') == 3_8)) then
      print *, 'expect failed'
      stop 1
    end if
  end subroutine test_example_3

  subroutine test_empty_string()
    implicit none
    if (.not. (lengthOfLongestSubstring('') == 0_8)) then
      print *, 'expect failed'
      stop 1
    end if
  end subroutine test_empty_string

end program main

File: examples/leetcode-out/scala/3/longest-substring-without-repeating-characters.scala
object Main {
    def lengthOfLongestSubstring(s: String): Int = {
        val n: Int = s.length
        var start: Int = 0
        var best: Int = 0
        var i: Int = 0
        val brk1 = new scala.util.control.Breaks
        brk1.breakable {
            while ((i < n)) {
                var j: Int = start
                val brk2 = new scala.util.control.Breaks
                brk2.breakable {
                    while ((j < i)) {
                        if ((_indexString(s, j) == _indexString(s, i))) {
                            start = (j + 1)
                            brk2.break()
                        }
                        j = (j + 1)
                    }
                }
                val length: Int = ((i - start) + 1)
                if ((length > best)) {
                    best = length
                }
                i = (i + 1)
            }
        }
        return best
    }
    
    def test_example_1(): Unit = {
        expect((lengthOfLongestSubstring("abcabcbb") == 3))
    }
    
    def test_example_2(): Unit = {
        expect((lengthOfLongestSubstring("bbbbb") == 1))
    }
    
    def test_example_3(): Unit = {
        expect((lengthOfLongestSubstring("pwwkew") == 3))
    }
    
    def test_empty_string(): Unit = {
        expect((lengthOfLongestSubstring("") == 0))
    }
    
    def main(args: Array[String]): Unit = {
        test_example_1()
        test_example_2()
        test_example_3()
        test_empty_string()
    }
}
def expect(cond: Boolean): Unit = {
        if (!cond) throw new RuntimeException("expect failed")
}

def _indexString(s: String, i: Int): String = {
        var idx = i
        val chars = s.toVector
        if (idx < 0) idx += chars.length
        if (idx < 0 || idx >= chars.length) throw new RuntimeException("index out of range")
        chars(idx).toString
}

